### ISSUE TITLE
`ansible-playbook --ask-become-pass` doesn't prompt properly

### ISSUE THREAD

**Issue by huyz** [xavierog/moulti#16]
If I pass `--ask-become-pass` to `ansible-playbook`, moulti doesn't seem to detect a prompt and give me an input field.

Then if I start typing the password anyway, the keyboard shortcuts will engage, e.g. `q` for the quit dialog

**Comment by xavierog** [comment]
Reproduced and confirmed.
Specifically, `--ask-become-pass` seems to request the password either directly through stdin/stdout/stderr/tty, or through an Ansible mechanism that remains to be identified.

**Comment by xavierog** [comment]
More specifically, the root cause is that `--ask-pass` and `--ask-become-pass` rely on [ansible.cli._get_secret()](https://github.com/ansible/ansible/blob/b5e0293645570f3f404ad1dbbe5f006956ada0df/lib/ansible/cli/__init__.py#L325), which in turn leverages [getpass.getpass()](https://docs.python.org/3/library/getpass.html), which interacts directly with `/dev/tty`.
Unless I am mistaken, Ansible does not even attempt to leverage its `Display` to get these passwords.

**Comment by xavierog** [comment]
Interestingly, there are only two uses of getpass.getpass() in Ansible:

1. in ansible.utils.Display.prompt(), which Moulti overrides
2. in ansible.cli.CLI._get_secret(), the cause of this bug.

It may be worth opening an issue in the Ansible repository suggesting to go through Display.prompt().

**Comment by xavierog** [comment]
>  Unless I am mistaken, Ansible does not even attempt to leverage its Display to get these passwords.

Stricto sensu, I am not mistaken... BUT ansible-playbook also provides options `--become-password-file` and `--connection-password-file`, and if an executable file is provided through these options, it gets executed and its standard output becomes the password, which is exactly why Moulti offers `moulti-askpass` in the first place.

I am going to experiment with these options. If they prove successful, Moulti could e.g. replace `--ask-become-pass` with `--become-password-file=/path/to/moulti-askpass`.

**Comment by xavierog** [comment]
@huyz I have pushed a fix to the `devel` branch. Feel free to test it.

One of the most important changes required to implement this fix is that `moulti run` may now change the command passed to it:
```
2024-08-25T01:13:17.085 exec: about to run ['ansible-playbook', '--ask-become-pass', '--ask-pass', 'playbook.yaml', '-C', '-D']
2024-08-25T01:13:17.087 exec: command changed to ['ansible-playbook', 'playbook.yaml', '-C', '-D']
2024-08-25T01:13:17.088 exec: ['ansible-playbook', 'playbook.yaml', '-C', '-D'] launched with PID 1234567
```

I also had to introduce two Ansible-specific variants of `moulti-askpass`: one for `--ask-become-pass`, one for `--ask-pass`:

```console
$ moulti run --print-env -- ansible-playbook --ask-become-pass --ask-pass playbook.yaml -C -D  | grep PASSWORD
ANSIBLE_CONNECTION_PASSWORD_FILE=/absolute/path/to/moulti-askpass-connection-password
ANSIBLE_BECOME_PASSWORD_FILE=/absolute/path/to/moulti-askpass-become-password
```

**Comment by xavierog** [comment]
Addendum: `--vault-password-file` relies on ansible.parsing.vault.PromptVaultSecret.ask_vault_passwords(), which leverages Display.prompt():
```python
vault_pass = display.prompt(prompt, private=True)
```
Theoretically, it should work... unless this is prompted too early, before Moulti's Display class is introduced... which is exactly what happens, of course.
I have pushed a second commit that deals with `--vault-password-file`.

We now have three Ansible-specific variants of `moulti-askpass`:
```
$ moulti run --print-env -- ansible-playbook --ask-become-pass --ask-pass --ask-vault-pass playbook.yaml -C -D  | grep PASSWORD
ANSIBLE_CONNECTION_PASSWORD_FILE=/absolute/path/to/moulti-askpass-connection-password
ANSIBLE_BECOME_PASSWORD_FILE=/absolute/path/to/moulti-askpass-become-password
ANSIBLE_VAULT_PASSWORD_FILE=/absolute/path/to/moulti-askpass-vault-client
```

**Comment by huyz** [comment]
I'm trying to run this, but the prompt is skipped entirely.

If I do `--print-env` I don't see any of the `*_PASSWORD_FILE` envvars.

Am I supposed to change anything else besides just doing git pull on the `devel` branch and rebuilding?

**Comment by xavierog** [comment]
> I'm trying to run this, but the prompt is skipped entirely.

...  which implies the command-line option was removed, that is a first step that works.

> If I do `--print-env` I don't see any of the `*_PASSWORD_FILE` envvars.

... which is possibly because the new variants of moulti-askpass-* are not available through $PATH.
 
> Am I supposed to change anything else besides just doing git pull on the `devel` branch and rebuilding?

Depends on what exact commands you call "rebuilding" here: it is necessary to pip(x) install so the new moulti-askpass variants are created in $PATH.

**Comment by huyz** [comment]
Oh I was just running `python setup.py install` in the git working tree.  I'll have to figure out how to do a complete "install" there then.  Or run `pip(x)` with the devel branch.

**Comment by xavierog** [comment]
> Fixed it by doing `pipx install git+https://github.com/xavierog/moulti.git@devel`.

Indeed, this is definitely the most straightforward AND reliable way to set up Moulti.

> Seems to work!

Fantastic!
 
> Btw (in case you wanted a more appropriate/user-friendly error message) if I hit CTRL-C during the password prompt, I get:
> 
> ```
> ERROR! The password script /Users/huyz/.local/pipx/venvs/moulti/bin/moulti-askpass-become-password returned an error (rc=1): b''
> ```

This is clearly an error message generated by Ansible, including that `b''` (meaning Ansible got 0 bytes from moulti-askpass) so I probably cannot improve that message. That being said, I am not sure to understand how you got it. Hitting Ctrl+c in Moulti should simply trigger the quit dialog.

**Comment by huyz** [comment]
> This is clearly an error message generated by Ansible, including that b'' (meaning Ansible got 0 bytes from moulti-askpass) so I probably cannot improve that message 

Not a problem.

> That being said, I am not sure to understand how you got it. Hitting Ctrl+c in Moulti should simply trigger the quit dialog.

I was being lazy; I meant hitting CTRL-c and then `t`

**Comment by xavierog** [comment]
> I was being lazy; I meant hitting CTRL-c and then `t`

Weird, I do not reproduce this behaviour. It might stem from the exact way macOS and Linux kill processes.

**Comment by xavierog** [comment]
This bugfix is now available as part of [Moulti 1.17.0](https://github.com/xavierog/moulti/releases/tag/v1.17.0).
